File: Command/CommandPattern.py
from abc import ABC, abstractmethod

class Invoker:  # Remote
    def __init__(self):
        self._commands = []

    def set_command(self, command):
        self._commands.append(command)

    def execute_commands(self):
        for command in self._commands:
            command.execute()
    
    def unexecute_commands(self):
        for command in self._commands:
            command.unexecute()
        

class Command(ABC): 
    @abstractmethod
    def execute(self):
        pass
    
    @abstractmethod
    def unexecute(self):
        pass


class ConcreteCommand(Command):
    def __init__(self, receiver):
        self._receiver = receiver

    def execute(self):
        self._receiver.action("Action 1")

    def unexecute(self):
        self._receiver.action("Action 2")


class Receiver: 
    def action(self, action):
        print(action)


#Client Code.
receiver = Receiver()
concrete_command = ConcreteCommand(receiver)
invoker = Invoker()
invoker.set_command(concrete_command)
invoker.execute_commands()
invoker.unexecute_commands()


"""
Action 1
Action 2
"""

File: Command/CommandPattern.java
import java.util.ArrayList;
import java.util.List;

// Invoker
class Invoker {
    private List<Command> commands = new ArrayList<>();

    public void setCommand(Command command) {
        commands.add(command);
    }

    public void executeCommands() {
        for (Command command : commands) {
            command.execute();
        }
    }

    public void unexecuteCommands() {
        for (Command command : commands) {
            command.unexecute();
        }
    }
}

// Command
abstract class Command {
    public abstract void execute();
    public abstract void unexecute();
}

// ConcreteCommand
class ConcreteCommand extends Command {
    private Receiver receiver;

    public ConcreteCommand(Receiver receiver) {
        this.receiver = receiver;
    }

    @Override
    public void execute() {
        receiver.action("Action 1");
    }

    @Override
    public void unexecute() {
        receiver.action("Action 2");
    }
}

// Receiver
class Receiver {
    public void action(String action) {
        System.out.println(action);
    }
}

// Client Code
public class CommandPattern {
    public static void main(String[] args) {
        Receiver receiver = new Receiver();
        ConcreteCommand concreteCommand = new ConcreteCommand(receiver);
        Invoker invoker = new Invoker();

        invoker.setCommand(concreteCommand);
        invoker.executeCommands();
        invoker.unexecuteCommands();
    }
}

/*
Action 1
Action 2
 */